### PR TITLE
Add a json-server compatible resource

### PR DESCRIPTION
**Pull request by morihladko**
This will add a json-server compatible resource. It extends the basic
`ModelResource` to use its query building, so it has to translate the
url query arguments from json-server and add some metadata to respond.

Making these changes required also:

 - reviving unit tests and creating some db fixtures. It was easier to
 write it against the predefined tests.

 - which fixed some bugs other bugs in `ModelResource`.

 - extending the resource router to support adding metadata and status
 codes into results.

 - adding other Access-Control-Allow-Headers

 - breaking down `ModelResource` to a base class and implementation
 class. I stopped right after that, because this will need a bigger
 refactor to split it into 3 parts - a REST router and HTTP responder,
 the REST command parser and the DB query executor.

 It also currently does not support PUT queries to replace the resource
 completely and extend parameters to embed parents.

### CODE REVIEW

File: tests/resource/model_resource.py [hunk 1]
```diff
@@ -3,97 +3,101 @@
 
 .. moduleauthor:: Peter Morihladko <morihladko@atteq.com>
 """
+from __future__ import (absolute_import, division,
+                        print_function, unicode_literals)
+from builtins import *
```
Comment (morihladko) [inline tests/resource/model_resource.py]: In [python-future.org](http://python-future.org/quickstart.html#if-you-are-writing-code-from-scratch) they recommend to use this, so I started to add in every new file.

File: tests/resource/resource_test_helper.py [hunk 2]
```diff
@@ -1,57 +1,116 @@
...
+    _app_model_class = UserAppModel
 
-    _app_model_class = DummyAppModel
+    id = Column('id', Integer, primary_key=True, nullable=False)
+    name = Column('name', String, nullable=False)
 
-    id = Column('id', INTEGER(), primary_key=True, nullable=False)
-    val = Column('val', VARCHAR(length=15), nullable=False)
+    def __repr__(self):
+        return '<User(id=\'{}\',name=\'{}\')>'.format(self.id, self.name)
```
Comment (morihladko) [inline tests/resource/resource_test_helper.py]: I renamed DummyModel to UserModel, it is easier to build upon something concrete e.g. some related model.

File: zsl/interface/resource.py [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+Helper types describing resource results for resource router.
+"""
+from __future__ import (absolute_import, division,
+                        print_function, unicode_literals)
+from builtins import *
+from collections import namedtuple
+
+result_fields = ['body', 'count', 'links', 'status', 'location', 'expose_headers']
+
+ResourceResult = namedtuple('ResourceResult', result_fields)
```
Comment (morihladko) [inline zsl/interface/resource.py]: I wanted to use something simple and expressive, so namedtuple.

File: zsl/interface/webservice/utils/response_headers.py [hunk 4]
```diff
@@ -20,13 +20,14 @@ def append_crossdomain(response, config=Injected):
 
     if config.get('ALLOW_ORIGIN'):
         response.headers['Access-Control-Allow-Origin'] = config.get('ALLOW_ORIGIN')
+
     response.headers['Access-Control-Allow-Methods'] = 'POST, GET, OPTIONS, PUT, DELETE'
-    response.headers['Access-Control-Allow-Headers'] = 'accept, origin, content-type'
+    response.headers['Access-Control-Allow-Headers'] = 'accept, origin, content-type, location, X-Total-Count, Link'
```
Comment (morihladko) [inline zsl/interface/webservice/utils/response_headers.py]: I don't know if the other headers should be added for all queries, but for now they are.

File: zsl/resource/model_resource.py [hunk 5]
```diff
@@ -169,11 +168,19 @@ def __init__(self, model_cls=None):
...
+        self._related_columns = [column.key for column in mapper.relationships]
+
+    def to_filter(self, query, arg):
+        return filter_from_url_arg(self.model_cls, query, arg)
+
+    def add_related(self, query, related):
+        return apply_related(self.model_cls, query, related)
+
+    def set_ordering(self, query, arg):
+        return order_from_url_arg(self.model_cls,  query, arg)
```
Comment (morihladko) [inline zsl/resource/model_resource.py]: I moved them from the constructor, there's no reason to create them in the constructor.

File: zsl/resource/resource_helper.py [hunk 6]
```diff
@@ -4,14 +4,16 @@
...
+from builtins import *
+from future.utils import viewitems, viewvalues
 
 from sqlalchemy.orm import class_mapper, joinedload
-from sqlalchemy import desc, asc
+from sqlalchemy import desc, asc, and_
 
 
-def filter_from_url_arg(model_cls, query, arg):
+def filter_from_url_arg(model_cls, query, arg, query_operator=and_):
```
Comment (morihladko) [inline zsl/resource/resource_helper.py]: Json-server joins filters with _or_ -_-.

File: zsl/resource/resource_helper.py [hunk 7]
```diff
@@ -22,7 +24,7 @@ def filter_from_url_arg(model_cls, query, arg):
     mapper = class_mapper(model_cls)
 
     exprs = []
-    joins = []
+    joins = set()
```
Comment (morihladko) [inline zsl/resource/resource_helper.py]: There was a bug where it tried to join same table multiple times.

File: zsl/resource/resource_helper.py [hunk 8]
```diff
@@ -32,10 +34,11 @@ def filter_from_url_arg(model_cls, query, arg):
 
         operator = None
         method = None
-        for op, m in viewitems(operator_to_method):
+        for op in operator_order:
```
Comment (morihladko) [inline zsl/resource/resource_helper.py]: There was a bug where in Py3 the dictionary order was not consistent.

File: tests/resource/model_resource.py (hunk 1 above)
Comment (babky-atteq) [inline tests/resource/model_resource.py]: The most important thing is to start unit tests on various Python environments Py2.7, Py3.5 and Py3.6. Yes, that is right. However, the old sources should be/are working as well. We'll see how we deal with the migration, when we set up the building on Travis.

File: zsl/interface/resource.py [hunk 9]
```diff
@@ -0,0 +1,15 @@
...
+ResourceResult.__new__.__defaults__ = (None,) * len(ResourceResult._fields)  # allow creating an empty result
```
Comment (babky-atteq) [inline zsl/interface/resource.py]: Just overlooked.

File: zsl/resource/json_server_resource.py [hunk 10]
```diff
@@ -0,0 +1,272 @@
...
+from .model_resource import ModelResource
+from zsl.interface.resource import ResourceResult
+from zsl.service.service import transactional
+from zsl.resource.resource_helper import filter_from_url_arg, flat_model, model_tree
+
+import http.client
+from flask import request
+
+
+def status_code(status):
```
Comment (babky-atteq) [inline zsl/resource/json_server_resource.py]: Elaborate which version is which code

Comment (babky-atteq) [inline zsl/resource/json_server_resource.py]: also this function should be module private and therefore is not so important

File: zsl/resource/json_server_resource.py [hunk 11]
```diff
@@ -0,0 +1,272 @@
...
+
+    if page > 1:
+        links['prev'] = page_url.replace(current_page, page_arg(page-1))
+
+    if end < count:
+        links['next'] = page_url.replace(current_page, page_arg(page+1))
+
+    if per_page < count:
+        links['first'] = page_url.replace(current_page, page_arg(1))
+        links['last'] = page_url.replace(current_page, page_arg(math.ceil(count / per_page)))
```
Comment (babky-atteq) [inline zsl/resource/json_server_resource.py]: i'd prefer (count + per_page - 1) // per_page computation
http://stackoverflow.com/questions/14822184/is-there-a-ceiling-equivalent-of-operator-in-python
plus if per_page and count are insanely large the above computation may fail

Comment (babky-atteq) [inline zsl/resource/json_server_resource.py]: at least page_count_method could be extracted from the code

File: zsl/resource/json_server_resource.py (hunk 10 above)
Comment (morihladko) [inline zsl/resource/json_server_resource.py]: done

**Pull request merged**